security-framework/src/certificate.rs

```diff
@@ -57,8 +86,152 @@ impl SecCertificate {
+                public_keysize_val_ptr as _,
+            );
+            if !have_vals {
+                CFRelease(public_key as _);
+                return Ok(None);
+            }
+            let hdr_bytes = get_asn1_header_bytes(public_key_type as _, public_keysize_val);
+            if hdr_bytes.len() == 0 {
+                CFRelease(public_key as _);
+                return Ok(None);
+            }
+
+            let public_key_data = SecKeyCopyExternalRepresentation(public_key, &mut error);
+            if public_key_data == ptr::null() {
+                CFRelease(public_key as _);
+                return Ok(None);
+            }
+
+            let key_data_len = CFDataGetLength(public_key_data) as usize;
+            let key_data_slice = std::slice::from_raw_parts(
+                CFDataGetBytePtr(public_key_data) as *const u8,
+                key_data_len,
+            );
+            let mut out = Vec::with_capacity(hdr_bytes.len() + key_data_len);
+            out.extend_from_slice(hdr_bytes);
+            out.extend_from_slice(key_data_slice);
+
+            CFRelease(public_key_data as _);
+            CFRelease(public_key as _);
+            Ok(Some(out))
+        }
+    }
+
+    #[cfg(any(feature = "OSX_10_12", target_os = "ios"))]
+    fn copy_public_key_from_certificate(&self) -> Result<SecKeyRef> {
+        unsafe {
+            // Create an X509 trust using the using the certificate
+            let mut trust: SecTrustRef = ptr::null_mut();
+            let policy: SecPolicyRef = SecPolicyCreateBasicX509();
+            cvt(SecTrustCreateWithCertificates(
+                self.as_concrete_TypeRef() as _,
+                policy as _,
+                &mut trust,
+            ))?;
+
+            // Get a public key reference for the certificate from the trust
+            let mut result: SecTrustResultType = 0;
+            cvt(SecTrustEvaluate(trust, &mut result))?;
+            let public_key = SecTrustCopyPublicKey(trust);
+            CFRelease(policy as _);
+            CFRelease(trust as _);
+            Ok(public_key)
+        }
+    }
 }
 
+#[cfg(any(feature = "OSX_10_12", target_os = "ios"))]
+fn get_asn1_header_bytes(pkt: CFStringRef, ksz: u32) -> &'static [u8] {
+    unsafe {
+        if CFStringCompare(pkt, kSecAttrKeyTypeRSA, 0) as i64 == 0 && ksz == 2048 {
+            return &RSA_2048_ASN1_HEADER;
+        }
+        if CFStringCompare(pkt, kSecAttrKeyTypeRSA, 0) as i64 == 0 && ksz == 4096 {
+            return &RSA_4096_ASN1_HEADER;
+        }
+        if CFStringCompare(pkt, kSecAttrKeyTypeECSECPrimeRandom, 0) as i64 == 0 && ksz == 256 {
+            return &EC_DSA_SECP_256_R1_ASN1_HEADER;
+        }
+        if CFStringCompare(pkt, kSecAttrKeyTypeECSECPrimeRandom, 0) as i64 == 0 && ksz == 384 {
+            return &EC_DSA_SECP_384_R1_ASN1_HEADER;
+        }
+    }
+    &[]
+}
+
+extern "C" {
+    #[cfg(any(feature = "OSX_10_12", target_os = "ios"))]
+    pub fn CFStringCompare(
```

Comment:
You don't need this. `CFString` implements `Eq`, so you can compare them with just `==`

Reply:
I'm using CFStringRef's due to comparing with constants kSecAttrKeyTypeRSA and kSecAttrKeyTypeECSECPrimeRandom. Can I turn those into CFString without allocating?

Reply:
I think you can. The `CFString` is a refcounted type, so a conversion from `Ref` to non-Ref should only need to change reference count:

https://docs.rs/core-foundation/0.6.3/core_foundation/base/trait.TCFType.html#tymethod.wrap_under_get_rule

Reply:
Also the eq is implemented as `self.as_CFType().eq(&other.as_CFType())`

Reply:
Ok switched to CFString now.